### PR TITLE
cilium: add probe for netkit for more user friendly error when not supported

### PR DESCRIPTION
**Pull request by borkmann**
(see commits)

### CODE REVIEW

**Comment by borkmann** [comment]
Cc @jrife for visibility

**Comment by jrife** [comment]
Hmm I wonder if it's worth checking if `SupportsScrub()` is true in this probe? We already have a warning if it's not, but I'm thinking users might miss that.

**Comment by borkmann** [comment]
> Hmm I wonder if it's worth checking if `SupportsScrub()` is true in this probe? We already have a warning if it's not, but I'm thinking users might miss that.

Hm, for the case of per-endpoint-routes we could actually bail out in that case.. for the non-per-ep-routes we don't need to warn.

**Comment by jrife** [comment]
That makes sense.

**Comment by borkmann** [comment]
> Edit: then again, this would basically make netkit unusable in the meantime since the agent would just not start up on most kernels people would be using today.

Ah okay, as I understand, without the patch users would at least have basic connectivity with per-ep-routes, right? If so then yes, we might keep the warning instead. Once the patch hits Linus tree on the next merge win, I'll ping Greg to backport to stable.

**Comment by jrife** [comment]
> Ah okay, as I understand, without the patch users would at least have basic connectivity with per-ep-routes, right?

Yes. It's only when you apply network policies that connectivity issues may arise in some cases.

**Review by jrife — APPROVED** [review]

File: daemon/cmd/daemon_main.go [hunk 1]
```diff
@@ -1273,6 +1273,9 @@ func initEnv(vp *viper.Viper) {
 		// before netkit and supporting legacy tc in this context does
 		// not make any sense whatsoever.
 		option.Config.EnableTCX = true
+		if err := probes.HaveNetkit(); err != nil {
+			log.Fatal("netkit devices need kernel 6.7.0 or newer")
```
Comment (qmonnet) [inline daemon/cmd/daemon_main.go]: Maybe add a note on `CONFIG_NETKIT` here as well?

**Review by qmonnet — APPROVED** [review]

File: daemon/cmd/daemon_main.go (hunk 1 above)
Comment (borkmann) [reply daemon/cmd/daemon_main.go]: Good idea, yeap.

**Comment by borkmann** [comment]
(CI was green, only changed error text as per @qmonnet's suggestion)

**Review by ldelossa — APPROVED** [review]
LGTM

**Review by bimmlerd — APPROVED** [review]

File: pkg/datapath/linux/probes/probes.go [hunk 2]
```diff
@@ -469,6 +469,51 @@ var HaveTCX = sync.OnceValue(func() error {
...
+	if err != nil {
+		return fmt.Errorf("create netns: %w", err)
+	}
+	defer ns.Close()
+
+	return ns.Do(func() error {
+		l, err := link.AttachNetkit(link.NetkitOptions{
+			Program:   prog,
+			Attach:    ebpf.AttachNetkitPrimary,
+			Interface: int(^uint(0) >> 1),
```
Comment (rgo3) [inline pkg/datapath/linux/probes/probes.go]: Nit:
```suggestion
			Interface: int(^uint32(0)),
```

**Review by rgo3 — APPROVED** [review]
LGTM

**Pull request merged**